---
title: "EpilepsyGenes"
author: "Enrico"
date: "19 September 2018"
output: html_document
---

# 0. Path Definition

Define the paths to different locations for easier calling during script

```{r Define paths, eval=FALSE, engine='bash'}

ssh -X ebazzicalupo@genomics-b.ebd.csic.es # these analyses will be conducted in server B because the annotated VCF file is located there

Epi_PATH=/home/ebazzicalupo/EpilepsyGenes # path to the project directory
Data_PATH=/home/ebazzicalupo/Data # path to data directory
LPRef_PATH=/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/lp23_plus_mt.fa # path to lynx pardinus reference genome file
VCF_PATH=/home/GRUPOS/grupolince/lynx_genomes_5x/VCFs_Dani/c_ll_lp_plus_h_ll_aafilled_SNPs_standard_filter.ann.vcf # path to annotated VCF file
gVCF_PATH=/home/GRUPOS/grupolince/lynx_genomes_5x/gVCFs # path to gVCFs
BDB_Path=/home/ebazzicalupo/Data/BLASTdatabase # path to BLAST database
DiscBDB_PATH=/home/ebazzicalupo/Data/Discarded_BLASTdatabase # path to Discarded Scaffold BLAST database
GATK=/opt/GATK-3.7/GenomeAnalysisTK.jar # path to GATK software

```


# 1. Search for Genes in Lynx pardinus Genome Annotations

Look for the genes names/abbreviations in the genome annotations files. First look in the Lynx pardinus genome annotations. If any are not found, look for them in Felis catus genome annotations [step 2], then in Homo sapiens genome annotations [step 3].

```{r Search in Lynx Annotations, eval=FALSE, engine='bash'}

## A file called EpilepsyGeneList.txt has been prepared with the gene names written for a boolean search through the grep command. Each line is a gene. The file has been located in the project directory (/home/ebazzicalupo/EpilepsyGenes).

## Genome annotation files for the three species have also been copied in the Data directory (/home/ebazzicalupo/Data). The Felis catus and Homo sapiens Genome Annotations have been downloaded from the ensembl website (https://www.ensembl.org/info/data/ftp/index.html).

cd $Epi_PATH

mkdir LynxResults # create directory to store the results for Lynx

GENES=$(cat EpilepsyGeneList.txt) # call all elements of EpilepsyGeneList.txt

IFS=$'\n' # define the end of a line as a separator for elements so that each ENTIRE line in EpilepsyGeneList.txt will be a search entry

# for each gene create a file containing results and move it to results directory
for i in ${GENES[@]}
  do
  grep -i -E "${i}" $Data_PATH/LYPA23C.all.fix.gff3 > ${i}
  mv ${i} LynxResults
done


```

Take the results and find the Parent Gene ID for each. Create a .txt file with the list of all Parent Gene IDs found.

```{r Refine Lynx Results, eval=FALSE, engine='bash'}

cd $Epi_PATH/LynxResults

LYNXREFINE=$(ls *) # list all results

mkdir ParentIDs # create a directory for storing the Parent Gene ID file

# for each result, cut the parent gene ID, sort it, eliminate duplicates and create a .Parent file with the Parent Gene IDs stored inside
for i in ${LYNXREFINE[@]}
  do cut -d ';' -f2 "${i}" | sort -u > "${i}".Parent
  mv "${i}".Parent ParentIDs
done

cd $Epi_PATH/LynxResults/ParentIDs

cat * | sort -u > ParentIDs.txt # put Parent Gene IDs from all the results in a single .txt file

mv ParentIDs.txt $Epi_PATH

```


# 2. Search in Felis cauts Genome Annotations for Genes not found in Lynx pardinus Genome Annotations. Extract corresponding sequence FASTA. Run BLAST to Lynx pardinus Reference Genome.

Repeat the search in Felis catus genome annotations, using as input the empty results from the Lynx search [step 2a]. This will give coordinates to specific regions in the Felis catus genome.
Use those coordinates to extract the corresponding sequence (.fa) from the Felis catus reference genome [step 2b], and BLAST it to the Lynx pardinus reference genome [step 2c]. Hit results from the BLAST will be analysed individually [step 4], to see if any genes in Lynx pardinus can be added to the Parent Gene ID list for further analyses [steps 5].

## 2a - Search in Cat annotations

```{r Search in Cat Annotations, eval=FALSE, engine='bash'}

cd $Epi_PATH

mkdir CatResults # create directory to store the results for Felis catus

CATGENES=$(wc -l LynxResults/* | grep ' 0 ' | cut -d '0' -f2 | sed 's/ //' | sed 's/LynxResults\///') # List empty results from Lynx search

IFS=$'\n' # define the end of a line as a separator for elements so that each ENTIRE line in PelageGeneList.txt will be a search entry

# for each gene create a file containing results and move it to results directory
for i in $CATGENES; do grep -i -E "${i}" $Data_PATH/Felis_catus.Felis_catus_8.0.92.gff3 > cat_"${i}".gff3; mv cat_"${i}".gff3 CatResults; done

```

## 2b - Generate Cat gene FASTA files for BLAST

```{r Generate Cat gene fasta files for BLAST, eval=FALSE, engine='bash'}

## Reference Genome FASTA files for Felis catus and Homo sapiens have been downloaded from the ensembl website (https://www.ensembl.org/info/data/ftp/index.html) and copied in the Data directory (/home/ebazzicalupo/Data).

cd $Epi_PATH/CatResults

CATREFINE=$(ls cat_*.gff3) # list results from grep search

# for each result from the grep search, create a gene.gff3 file with the non-gene (e.g. mRNA, CDS) strings removed
for i in ${CATREFINE[@]}
  do
  awk -v "key=gene" '$3 == key {print($0)}' ${i} > ${i/.gff3}.gene.gff3
done


CATcoords=$(ls cat_*.gene.gff3) # list gene.gff3 files

# generate FASTA file for each gene.gff3 coordinates file
for i in ${CATcoords[@]}
  do
  bedtools getfasta -fo "${i/.gene.gff3}".fa -fi $Data_PATH/Felis_catus.Felis_catus_8.0.dna_sm.toplevel.fa -bed ${i}
done

```

## 2c - BLAST cat genes to Lynx genome

```{r  BLAST cat genes to Lynx genome, eval=FALSE, engine='bash'}

## First create a BLASTdatabase for the Lynx pardinus FASTA file

cd $Data_PATH

mkdir BLASTdatabase # create a directory for the BLAST database

BDB_Path=/home/ebazzicalupo/Data/BLASTdatabase # define path to BLAST database

scp $LPRef_PATH $BDB_Path # copy reference genome to the BLASTdatabase directory

# create a BLAST database in the BLASTdatabase directory
makeblastdb -in $BDB/lp23_plus_mt.fa -parse_seqids -dbtype nucl

## BLAST each Cat Gene FASTA to the Lynx pardinus Reference Genome and BLAST database. Look at BLAST manual for output options.

cd $Epi_PATH/CatResults

IFS=$'\n'

CBLASTGENES=$(ls cat_*.fa) # list Cat Gene FASTA files

for GENE in ${CBLASTGENES[@]}
  do
  echo "$GENE starting BLAST"
  blastn -query $Epi_PATH/CatResults/${GENE} -db $BDB_Path/lp23_plus_mt.fa -outfmt "6 qseqid sseqid sstart send length qcovhsp evalue" -out ${GENE/.fa/}.BLASTresults
  echo "$GENE BLAST finished"
done

## Move the BLASTresults to a new directory

mkdir BLASTresults

mv *.BLASTresults BLASTresults

```

# 3. Search in Homo sapiens Genome Annotations for Genes not found in Lynx pardinus and Felis catus Genome Annotations. Extract corresponding sequence FASTA. Run BLAST to Lynx pardinus Reference Genome.

Repeat the search in Homo Sapiens genome annotations, using as input the empty results from the Felis catus search [step 3a]. This will give coordinates to specific regions in the Homo sapiens.
Use those coordinates to extract the corresponding sequence (.fa) from the Homo sapiens reference genome [step 3b], and BLAST it to the Lynx pardinus reference genome [step 3c]. Hit results from the BLAST will be analysed individually, to see if any genes in Lynx pardinus can be added to the Parent Gene ID list for further analyses [steps XX].

## 3a - Search in Human annotations

```{r Search in Human Annotations, eval=FALSE, engine='bash'}

cd $Epi_PATH

mkdir HumanResults # create directory to store the results for Homo sapiens

HUMANGENES=$(wc -l CatResults/*.gene.gff3 | grep ' 0 ' | cut -d '0' -f2 | sed 's/CatResults\/cat_//g' | sed 's/ //') # List empty results from Cat search

IFS=$'\n'


# for each gene create a gff3 file containing results and move it to HumanResults folder
for i in ${HUMANGENES[@]}
  do
  grep -i -E "${i/.gene.gff3/}" $Data_PATH/Homo_sapiens.GRCh38.92.gff3 > human_${i/.gene.gff3/}.gff3
  mv human_${i/.gene.gff3/}.gff3 HumanResults
done

```

## 3b - Generate Human gene FASTA files for BLAST

```{r Generate Human gene fasta files for BLAST, eval=FALSE, engine='bash'}

## Reference Genome FASTA files for Felis catus and Homo sapiens have been downloaded from the ensembl website (https://www.ensembl.org/info/data/ftp/index.html) and copied in the Data directory (/home/ebazzicalupo/Data).

cd $Epi_PATH/HumanResults

HUMANREFINE=$(ls human_*.gff3) # list results from grep search

IFS=$'\n'

# for each result from the grep search, create a gene.gff3 file with the non-gene (e.g. mRNA, CDS) strings removed
for i in ${HUMANREFINE[@]}
  do
  awk -v "key=gene" '$3 == key {print($0)}' ${i} > ${i/.gff3}.gene.gff3
done

HUMANcoords=$(ls human_*.gene.gff3) # list gene.gff3 files

# generate FASTA file for each gene.gff3 coordinates file
for i in ${HUMANcoords[@]}
  do
  bedtools getfasta -fo "${i/.gene.gff3}".fa -fi $Data_PATH/Homo_sapiens.GRCh38.dna_sm.toplevel.fa -bed ${i}
done

```

## 3c - BLAST human genes to Lynx genome

```{r  BLAST human genes to Lynx genome, eval=FALSE, engine='bash'}

## BLAST each Human Gene FASTA to the Lynx pardinus Reference Genome and BLAST database. Look at BLAST manual for output options.

cd $Epi_PATH/HumanResults

IFS=$'\n'

HBLASTGENES=$(ls human_*.fa) # list Human Gene FASTA files

for GENE in ${HBLASTGENES[@]}
  do
  echo "$GENE starting BLAST"
  blastn -query $Epi_PATH/HumanResults/${GENE} -db $BDB_Path/lp23_plus_mt.fa -outfmt "6 qseqid sseqid sstart send length qcovhsp evalue" -out ${GENE/.fa/}.BLASTresults
  echo "$GENE BLAST finished"
done

## Move the BLASTresults to a new directory

mkdir BLASTresults

mv *.BLASTresults BLASTresults

```

# 4. Generate a VCF with the list of SNPs of our genes of interest.

We need to extract from the resulting VCF file the positions that fall within the genes we are interested in. Using the grep command we are able to extract not only the variants that fall within the gene, but also intergenic, upstream or downstream regions which might be of interest. The information will be extracted from the unfiltered annotated VCF which includes both L.lynx and L.pardinus

## 4a - Look for (grep) ParentIDs in VCF file

```{r  Grep ParentIDs from VCF, eval=FALSE, engine='bash'}

## Prepare ParentIDs file for grep
cd $Epi_PATH

# The ParentIDs list contains "Parent=" before each ID, interfering with the search. 
sed 's/Parent=//g' ParentIDs.txt > ParentIDs_onlyIDs.txt

# Add the '#' sign to the list in order to copy the header part of the VCF too
nano ParentIDs_onlyIDs.txt


## Grep with annotated VCF file including both species

cd $Epi_PATH
mkdir VCF
cd $Epi_PATH/VCF

# A copy of the VCF file including both species has been copied in $Epi_PATH/VCF

scp /home/GRUPOS/grupolince/lynx_genomes_5x/VCFs_Dani/annotation/c_ll_lp_plus_h_ll_polarized.lr_ann.vcf $Epi_PATH/VCF

# Grep all elements in the ParentIDs list from the VCF

grep -f $Epi_PATH/ParentIDs_onlyIDs.txt $Epi_PATH/VCF/c_ll_lp_plus_h_ll_polarized.lr_ann.vcf > $Epi_PATH/VCF/c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep.vcf

# Sort SNPs
sort -k1,2 $Epi_PATH/VCF/c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep.vcf > tmp && mv tmp $Epi_PATH/VCF/c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep.vcf

```

# 5. Create a table with the useful SNPs information and annotations from VCF file

A table is created with the information regarding the SNPs using the VCF file contaning information for both species (see step 4a). This version tries to simplify the final result. It has the advantage of having a single column for SNPEFF annotations, which with a larger dataset are also more numerous. There is a unique column for all of the CUSTOM annotations containing the information regarding the "feature" (region where the SNP falls). All of the statistics in 5a version have been removed for increased clarity, leaving only the information regarding the Ancestral Allel and the Annotation.

Divided in sections for easier consultation.

```{r Create a table with the useful SNPs - Section 1, eval=FALSE, engine='bash'}

## Section 1: Preprocess and divide in NoAnnotations, SNPEFF annotations and CUSTOM annotations

# remove header and information about individuals
grep -v '#' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep.vcf | cut -f1-8 > c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals.vcf

# replace blank spaces in annotations with "na"
sed 's/||/|na|/g' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals.vcf | sed 's/||/|na|/g' > c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas.vcf

# separate annotations from rest of VCF

awk -F ';ANN='  '{print $1}' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas.vcf > c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_noann.vcf

awk -F 'ANN='  '{print $2}' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas.vcf > c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly.vcf

# Remove Custom annotations - SNPEFF annotations ONLY are LEFT
sed 's/,.*|custom|MODIFIER|na|na|CUSTOM&LYPA23C|.*|na|na|na|na|na|//g' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly.vcf > c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff.vcf

# Remove Loss of Function (LOF) from SNPEFF only
sed 's/;LOF.*)//g' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff.vcf > c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof.vcf

# get all lines with LOF and their line number - LOF ONLY are LEFT
grep -n -o ";LOF.*)" c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff.vcf > c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_lofonly.vcf

# get all lines with custom annotations and the corresponding line number - CUSTOM annotations ONLY are LEFT
grep -n -o ",.|custom|MODIFIER|na|na|CUSTOM&LYPA23C|.*|na|na|na|na|na|" c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly.vcf > c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_customonly.vcf

```

## SNPEFF

```{r Create a table with the useful SNPs - Section 2, eval=FALSE, engine='bash'}

## Section 2: Generate a SNPEFF columns file with SNPEFF annotations decomposed for each line

# Make all of SNPEFF annotations have same number of fields (16 normally, 15 if it's the last annotation on the line without warning messages)

sed 's/|,/|na,/g' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof.vcf | sed 's/,/|,/g' > c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf

# count maximum number of SNPEFF annotations

max=$(awk -v FS=',' 'max < NF { max = NF } END { print max }' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf ) && echo $max # count number of columns to include -> in this case 16.

# (column 2) get ANNOTATION information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the second)
cut -d '|' -f2,18,34,50,66,82,98,114,130,146,162,178,194,210,226,242 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Snpeff_annotation

# (column 3) get IMPACT information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the third)
cut -d '|' -f3,19,35,51,67,83,99,115,131,147,163,179,195,211,227,243 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Snpeff_impact

# (column 5) get GENE ID information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the fifth)
cut -d '|' -f5,21,37,53,69,85,101,117,133,149,165,181,197,213,229,245 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Snpeff_geneID

# (column 6) get FEATURE TYPE information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the sixth)
cut -d '|' -f6,22,38,54,70,86,102,118,134,150,166,182,198,214,230,246 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Snpeff_featuretype

# (column 7) get FEATURE ID information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the seventh)
cut -d '|' -f7,23,39,55,71,87,103,119,135,151,167,183,199,215,231,247 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Snpeff_featureID

# (column 8) get BIOTYPE information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the eitghth)
cut -d '|' -f8,24,40,56,72,88,104,120,136,152,168,184,200,216,232,248 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n'  > Snpeff_biotype

# (column 9) get intron/exon RANK / TOTAL information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the ninth)
cut -d '|' -f9,25,41,57,73,89,105,121,137,153,169,185,201,217,233,249 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Snpeff_rank

# (column 10) get HGVS variant information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the tenth)
cut -d '|' -f10,26,42,58,74,90,106,122,138,154,170,186,202,218,234,250 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Snpeff_hgvsvar

# (column 11) get HGVSprotein variant information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the eleventh)
cut -d '|' -f11,27,43,59,75,91,107,123,139,155,171,187,203,219,235,251 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Snpeff_protvar

# (column 12) get cDNA POS/LENGTH information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the twelth)
cut -d '|' -f12,28,44,60,76,92,108,124,140,156,172,188,204,220,236,252 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Snpeff_cdnapos

# (column 13) get CDS POS/LENGTH information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the thirteenth)
cut -d '|' -f13,29,45,61,77,93,109,125,141,157,173,189,205,221,237,253 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Snpeff_cdspos

# (column 14) get PROTEIN POS/LENGTH information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the fourteenth)
cut -d '|' -f14,30,46,62,78,94,110,126,142,158,174,190,206,222,238,254 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Snpeff_protpos

# (column 15) get DISTANCE TO FEATURE information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the fifteenth)
cut -d '|' -f15,31,47,63,79,95,111,127,143,159,175,191,207,223,239,255 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Snpeff_featdist

# (column 16) get ERRORS information - Having 16 fields of snpeff annotations we get every 16th column starting from the column of interest (in this case the sixteenth)
cut -d '|' -f16,32,48,64,80,96,112,128,144,160,176,192,208,224,240,256 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nolof_equalfields.vcf | sed 's/|na|//g' | sed 's/na|//g' | sed 's/|na//g' | sed 's/na//g' | sed -e 's/^$/NA/' | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Snpeff_errors

# paste everything together
paste Snpeff_annotation Snpeff_impact Snpeff_geneID Snpeff_featuretype Snpeff_featureID Snpeff_biotype Snpeff_rank Snpeff_hgvsvar Snpeff_protvar Snpeff_cdnapos Snpeff_cdspos Snpeff_protpos Snpeff_featdist Snpeff_errors > SNPEFF_COMPLETE

# remove intermediate files
rm Snpeff_*

```

## CUSTOM

```{r Create a table with the useful SNPs - Section 3, eval=FALSE, engine='bash'}

## Section 3: Condense all of Custom Annotations into one column and add blank lines for positions missing Custom annotations

# generate a file with the corresponding line number for the custom annotations - CUSTOM ONLY
cut -d ':' -f1 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_customonly.vcf > aa_customlines

# generate a file with all the numbers from 1 to the maximum line number (total number of lines in the original file)
TotalNumberOfLines=$(cat c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas.vcf | wc -l)

seq $TotalNumberOfLines > aa_totlines

# subtract the custom line numbers from the total number of lines to generate a file listing the missing lines (lines with no custom annotation)
comm aa_customlines aa_totlines | cut -f2 | sed '/^\s*$/d' > aa_nocustomlines # NOT WORKING ON SERVER -> copied aa_customlines and aa_totlines to my laptop, executed this command there and then copied aa_nocustomlines back to server - ACTUALLY it probably worked, done it on laptop just to be sure

comm aa_customlines aa_totlines > aa_nocustomlines

# add a white line in the CUSTOM ONLY annotations file in the missing lines - CAREFUL AS IT OVERWRITES - CHECK IF DONE ALREADY / OTHER POSSIBLE PROBLEMS - if problems are found better re-generate c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_customonly.vcf from scratch -> VERY SLOW better run it on a screen
NoCustomLines=$(cat aa_nocustomlines)

for i in $NoCustomLines
  do
  awk -v n=${i} -v s="" 'NR == n {print s} {print}' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_customonly.vcf > tmp && mv tmp c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_customonly.vcf
done

# remove useless files used for previous steps
rm aa_*

# remove comma and line number before pasting the two tables together - CUSTOM ONLY
cut -d ':' -f2 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_customonly.vcf | sed 's/,//' > c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_customonly_blanklines.vcf

# condense all information in unique custom column: first count the maximum number of custom annotation columns (max); then cut columns 7, 22, 37, 52, 67 ecc. (always add 15 to the previous starting from 7) a number of times equal to max (to catch all) - CUSTOM ONLY

max=$(awk -v FS=',' 'max < NF { max = NF } END { print max }' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_customonly_blanklines.vcf) && echo $max # count number of columns to include

cut -d '|' -f7,22,37,52,67,82,97,112,127,142,157,172 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_customonly_blanklines.vcf > c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_customonly_blanklines_condensed.vcf # in this case max=12 so columns 7,22(7+15),37(7+30),52(7+45),67,82,97,112,127,142,157,172

# insert EMPTY in empty lines and remove repetitions CUSTOM ONLY
sed -e 's/^$/EMPTY/' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_customonly_blanklines_condensed.vcf | sed 's/|/ /g' | sed ':s;s/\(\<\S*\>\)\(.*\)\<\1\>/\1\2/g;ts;s/  */ /g' | sed 's/ /|/g' | tr '\n' ' ' | sed 's/| / /g' | tr ' ' '\n' > Custom_ONLY

```

## LOF

```{r Create a table with the useful SNPs - Section 4, eval=FALSE, engine='bash'}

## Section 4: Create column for Loss of Function Annotations (LOF)

# generate a file with the corresponding line number for the LOF annotations - LOF ONLY
cut -d ':' -f1 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_lofonly.vcf > aa_loflines

# generate a file with all the numbers from 1 to the maximum line number (total number of lines in the original file)
TotalNumberOfLines=$(cat c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas.vcf | wc -l)

seq $TotalNumberOfLines > aa_totlines

# subtract the custom line numbers from the total number of lines to generate a file listing the missing lines (lines without LOF annotations)
grep -F -x -v -f aa_loflines aa_totlines > aa_noloflines

# add a white line in the LOF ONLY annotations file in the missing lines - CAREFUL AS IT OVERWRITES - CHECK IF DONE ALREADY / OTHER POSSIBLE PROBLEMS - if problems are found better re-generate c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_lofonly.vcf from scratch -> VERY SLOW better run it on a screen
NoLOFLines=$(cat aa_noloflines)
echo "" >> c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_lofonly.vcf # add one empty line at the end of the file so that the script will fill all empty spaces

for i in $NoLOFLines
  do
  awk -v n=${i} -v s="" 'NR == n {print s} {print}' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_lofonly.vcf > tmp && mv tmp c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_lofonly.vcf
done

sed '$d' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_lofonly.vcf > tmp && mv tmp c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_lofonly.vcf # remove the empty line previously added

# remove comma, remove line number, add NA to empty columns before pasting the two tables together - CUSTOM ONLY
cut -d ':' -f2 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_lofonly.vcf | sed 's/,//' | sed -e 's/^$/NA/' > LOF_annotations

```

## PASTE

```{r Create a table with the useful SNPs - Section 5, eval=FALSE, engine='bash'}

## Section 5: Paste all tables together


# paste SNPEFF ONLY (c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_snpeff_nocomma.vcf) and CUSTOM ONLY (c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_annonly_customonly_blanklines_condensed.vcf) tables together

paste -d '\t' Custom_ONLY SNPEFF_COMPLETE LOF_annotations > ALL_Annotations


# BEFORE PASTING THE NO ANNOTATIONS FILE NEEDS MODIFICATIONS - keep only AA information

cut -d ';' -f1 c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_noann.vcf > c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_noann_aainfoonly.vcf


# paste annotations to table without no annotations table

paste -d '\t' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_noIndividuals_nas_noann_aainfoonly.vcf ALL_Annotations > c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_squaredtable.vcf


# Final adjustments to table

# Remove "AA=" before the AA

sed 's/AA=//' c_ll_lp_plus_h_ll_polarized.lr_ann.ParentIDsGrep_squaredtable.vcf > EpilepsyGenesTable.vcf

# HEADER
SCAFFOLD	POS	ID	REF	ALT	QUAL	FILTER	AA	CUSTOM	SNPEFF	IMPACT	GENE_ID	FEATURE	FEATURE_ID	BIOTYPE	RANK	VARIANT	P_VARIANT	CDNA_POS	CDS_POS	PROT_POS	DISTANCE	ERRORS

```
